src/backend/distributed/planner/distributed_planner.c

```diff
@@ -238,9 +263,12 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 			result = CreateDistributedPlannedStmt(planId, result, originalQuery, parse,
 												  boundParams, plannerRestrictionContext);
 
-			setPartitionedTablesInherited = true;
-			AdjustPartitioningForDistributedPlanning(rangeTableList,
-													 setPartitionedTablesInherited);
+			if (!fastPathRouterQuery)
+			{
```

Comment:
There are a lot of fastPathQuery conditions in this function. Can we write a different variant of distributed_planner that excludes all the non-fast-path bits?

Reply:
That sounds a big change with some edge cases like handling `PlannerLevel` or `plannerRestrictionContext` etc.

One simple approach could be to change the following and get rid of the other `if` checks
```C
if (needsDistributedPlanning && FastPathRouterQuery(parse, &distributionKeyValue))
{
	fastPathRouterQuery = true;
 
    /*  we don't need any further checks on the rangeTableList */
    rangeTableList =  NIL;
}
```

And, since `rangeTableList` becomes `NIL`, we would avoid many checks/loops. 

If  we  can find a  better name `rangeTableList` to make it explicit, it'd be even better.

How does that sound?

Reply:
My main concern is the understandability of this function, which is already hard to follow. Setting `rangeTableList = NIL` does not help understandability.

Reply:
My concern on separating the code-path for fast-path is that we'd end-up with handling the same things twice like the hacks for prepared statements, or maintaining `PlannerLevel` needs to be doubled. Or,  we're now putting some information to restriction context in both cases, which we may need to change because in fast-path Push/Pop restriction context doesn't actually make sense.

One approach I  can think of, we can have  a function like `PrepareForDistributedPlanning()` which captures the first ~50 lines of `distributed_planner`  and `PostDistributedPlanning()` to clean-up some of the logic.

How does that sound to you?

(And, one another thing  that I  don't like about this function is that `TryToDelegateFunctionCall` is called in `!needsDistributedPlanning`, which  adds complexity in  my point  of  view.)

Reply:
I refactored this now to have a clearer separation between the 3 planning logic: 
- distributed fast path
- distributed non fast path
- non distributed plan

I think that makes much easier to understand what is happening.